### PR TITLE
OAuthPrompt timeout applies to all auth related events and invokes.

### PR DESCRIPTION
**Pull request by EricDahlvang**
Fixes https://github.com/microsoft/botbuilder-dotnet/issues/3627

This PR also prioritizes the timeout check before RecognizeTokenAsync, since RecognizeTokenAsync has side affects.

### CODE REVIEW

File: libraries/Microsoft.Bot.Builder.Dialogs/Prompts/OAuthPrompt.cs [hunk 1]
```diff
@@ -303,6 +307,7 @@ private static bool ChannelSupportsOAuthCard(string channelId)
                 case Channels.Cortana:
                 case Channels.Skype:
                 case Channels.Skypeforbusiness:
+                case Channels.Msteams:
                     return false;
             }
```
Comment (tomlm) [inline libraries/Microsoft.Bot.Builder.Dialogs/Prompts/OAuthPrompt.cs:313]: What?  I thought all of our channels do, in particular WebChat?

File: libraries/Microsoft.Bot.Builder.Dialogs/Prompts/OAuthPrompt.cs [hunk 2]
```diff
@@ -303,6 +307,7 @@ private static bool ChannelSupportsOAuthCard(string channelId)
                 case Channels.Cortana:
                 case Channels.Skype:
                 case Channels.Skypeforbusiness:
+                case Channels.Msteams:
                     return false;
```
Comment (tomlm) [inline libraries/Microsoft.Bot.Builder.Dialogs/Prompts/OAuthPrompt.cs:311]: AFAIK Teams supports OAuthCard!?!?

Comment (cleemullins) [reply libraries/Microsoft.Bot.Builder.Dialogs/Prompts/OAuthPrompt.cs:311]: @tomlm I believe the SSO flow uses the OAuth Prompt, rather than the OAuth card. I know Carlos / Eric / Swagat have gone pretty deep here, and this is where they settled. @EricDahlvang  Can you elaborate a bit?

File: libraries/Microsoft.Bot.Builder.Dialogs/Prompts/OAuthPrompt.cs [hunk 3]
```diff
@@ -179,21 +179,25 @@ public override async Task<DialogTurnResult> ContinueDialogAsync(DialogContext d
...
 
-            // Recognize token
-            var recognized = await RecognizeTokenAsync(dc, cancellationToken).ConfigureAwait(false);
-
             // Check for timeout
             var state = dc.ActiveDialog.State;
             var expires = (DateTime)state[PersistedExpires];
             var isMessage = dc.Context.Activity.Type == ActivityTypes.Message;
-            var hasTimedOut = isMessage && DateTime.Compare(DateTime.Now, expires) > 0;
+            var isTimeoutActivityType = isMessage
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder.Dialogs/Prompts/OAuthPrompt.cs:186]: Consider adding a bit detailed comment here around what's going on and why this is here. I happen to remember now from the conversations yesterday, but months from now the details will be gone. Having a detailed comment around "Why is this here?" would be great.

Comment (EricDahlvang) [reply libraries/Microsoft.Bot.Builder.Dialogs/Prompts/OAuthPrompt.cs:189]: Added descriptive comment

File: libraries/Microsoft.Bot.Builder.Dialogs/Prompts/OAuthPrompt.cs (hunk 2 above)
Comment (EricDahlvang) [reply libraries/Microsoft.Bot.Builder.Dialogs/Prompts/OAuthPrompt.cs]: Tom is correct... this was in unintentional change.  (Teams does now support OAuthCards, as long as the SignInLink is present)

NOTE: after seeing Tom's comment here, i've also removed Channels.Msteams from **OAuthInput**

**Comment by cleemullins** [comment]
@carlosscastro We beleive this is good to merge. Tom has looked at it, and Eric confident. Just because OAuth and SSO are such nuanced areas, can you take a quick peek before we merge?

File: libraries/Microsoft.Bot.Builder.Dialogs/Prompts/OAuthPrompt.cs [hunk 4]
```diff
@@ -179,21 +179,28 @@ public override async Task<DialogTurnResult> ContinueDialogAsync(DialogContext d
...
             var isMessage = dc.Context.Activity.Type == ActivityTypes.Message;
-            var hasTimedOut = isMessage && DateTime.Compare(DateTime.Now, expires) > 0;
+
+            // If the incoming Activity is a message, or an Activity Type normally handled by OAuthPrompt,
+            // check to see if this OAuthPrompt Expiration has elapsed, and end the dialog if so.
+            var isTimeoutActivityType = isMessage
+                            || IsTokenResponseEvent(dc.Context)
+                            || IsTeamsVerificationInvoke(dc.Context)
+                            || IsTokenExchangeRequestInvoke(dc.Context);
+            var hasTimedOut = isTimeoutActivityType && DateTime.Compare(DateTime.Now, expires) > 0;
```
Comment (carlosscastro) [inline libraries/Microsoft.Bot.Builder.Dialogs/Prompts/OAuthPrompt.cs:193]: DateTime.UtcNow

Comment (carlosscastro) [reply libraries/Microsoft.Bot.Builder.Dialogs/Prompts/OAuthPrompt.cs:193]: Also, it's more readable if you compare dates with the operator IMO instead of DateTime.Compare, i.e. date1 > date2
